addons/binding/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/WebSocketConnection.java

```diff
@@ -0,0 +1,592 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.amazonechocontrol.internal;
+
+import java.io.IOException;
+import java.net.HttpCookie;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.List;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.UUID;
+import java.util.concurrent.ThreadLocalRandom;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.util.ssl.SslContextFactory;
+import org.eclipse.jetty.websocket.api.Session;
+import org.eclipse.jetty.websocket.api.WebSocketListener;
+import org.eclipse.jetty.websocket.client.ClientUpgradeRequest;
+import org.eclipse.jetty.websocket.client.WebSocketClient;
+import org.openhab.binding.amazonechocontrol.internal.jsons.JsonPushCommand;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The {@link WebSocketConnection} encapsulate the Web Socket connection to the amazon server.
+ * The code is based on
+ * https://github.com/Apollon77/alexa-remote/blob/master/alexa-wsmqtt.js
+ *
+ * @author Michael Geramb - Initial contribution
+ * @author Ingo Fischer - (https://github.com/Apollon77/alexa-remote/blob/master/alexa-wsmqtt.js)
+ */
+@NonNullByDefault
+public class WebSocketConnection {
+    private final Logger logger = LoggerFactory.getLogger(WebSocketConnection.class);
+    private final Gson gson = new Gson();
+    WebSocketClient webSocketClient;
+    @Nullable
+    Session session;
+    @Nullable
+    Timer pingTimer;
+    @Nullable
+    Timer pongTimeoutTimer;
+    Listener listener;
+    boolean closed;
+    IWebSocketCommandHandler webSocketCommandHandler;
+
+    public WebSocketConnection(String amazonSite, List<HttpCookie> sessionCookies,
+            IWebSocketCommandHandler webSocketCommandHandler) throws IOException {
+        this.webSocketCommandHandler = webSocketCommandHandler;
+        listener = new Listener();
+        SslContextFactory sslContextFactory = new SslContextFactory();
+        webSocketClient = new WebSocketClient(sslContextFactory);
```

Comment:
Sorry, I meant the [WebSocketFactory](https://github.com/eclipse/smarthome/blob/master/bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/WebSocketFactory.java) - it is an OSGi service that can be used to retrieve a shared client.

Reply:
What is the job of this factory, is there any advantage? I can not see any example, the bosesoundtouch, loxonevalloxmv and innogysmarthome creating there own WebSocketClients. The socket must be open of the whole lifetime of the thing. And I have a problem with the IDE, it does not find the WebSocketFactory class.

Reply:
It has only recently been introduced. It avoids an abundance of threads as Jetty opens a lot for every client instance (while a single instance is capable of many connections without issues).
Check the SonyAudio binding, which already uses the factory.

> And I have a problem with the IDE, it does not find the WebSocketFactory class.

Make sure to reload your target platform, the current version should have it (at least since mid of November).